src/main/java/com/conveyal/gtfs/stats/PatternStats.java

```diff
@@ -1,7 +1,176 @@
+                speeds.add(speed);
+            }
+        }
+
+        if (speeds.isEmpty()) return Double.NaN;
+
+        return speeds.sum() / speeds.size();
+    }
+
+    public LocalTime getStartTimeForTrips (Collection<Trip> trips) {
+        int earliestDeparture = Integer.MAX_VALUE;
+
+        for (Trip trip : trips) {
+            StopTime st;
+            try {
+                // use interpolated times in case our common stop is not a time point
+                st = StreamSupport.stream(feed.getInterpolatedStopTimesForTrip(trip.trip_id).spliterator(), false)
+                        .findFirst()
+                        .orElse(null);
+            } catch (GTFSFeed.FirstAndLastStopsDoNotHaveTimes e) {
+                return null;
+            }
+
+            // shouldn't actually encounter any departures after midnight, but skip any departures that do
+            if (st.departure_time > 86399) continue;
+
+            if (st.departure_time <= earliestDeparture) {
+                earliestDeparture = st.departure_time;
+            }
+        }
+        return LocalTime.ofSecondOfDay(earliestDeparture);
+    }
+
+    public LocalTime getEndTimeForTrips (Collection<Trip> trips) {
+        int latestArrival = Integer.MIN_VALUE;
+
+        for (Trip trip : trips) {
+            StopTime st;
+            try {
+                // use interpolated times in case our common stop is not a time point
+                st = StreamSupport.stream(feed.getInterpolatedStopTimesForTrip(trip.trip_id).spliterator(), false)
+                        .reduce((a, b) -> b)
+                        .orElse(null);
+            } catch (GTFSFeed.FirstAndLastStopsDoNotHaveTimes e) {
+                return null;
+            }
+
+            if (st.arrival_time >= latestArrival) {
+                latestArrival = st.arrival_time;
+            }
+        }
+
+        // return end time as 2:00 am if last arrival occurs after midnight
+        return LocalTime.ofSecondOfDay(latestArrival % 86399);
+    }
+
+    /** Get total revenue time (in seconds) for set of trips. */
+    public int getTotalRevenueTimeForTrips (Collection<Trip> trips) {
+        TIntList times = new TIntArrayList();
+        for (Trip trip : trips) {
+            StopTime first;
+            StopTime last;
+            Spliterator<StopTime> interpolated = null;
+            try {
+                interpolated = feed.getInterpolatedStopTimesForTrip(trip.trip_id).spliterator();
+            } catch (GTFSFeed.FirstAndLastStopsDoNotHaveTimes firstAndLastStopsDoNotHaveTimes) {
+                return -1;
+            }
+
+            first = StreamSupport.stream(interpolated, false)
+                    .findFirst()
+                    .orElse(null);
+
+            last = StreamSupport.stream(interpolated, false)
+                    .reduce((a, b) -> b)
+                    .orElse(null);
+            int time = last.departure_time - first.arrival_time;
```

Comment:
Shouldn't this be `last.arrival_time - first.departure_time`? I wouldn't consider layover at first and last stops to be revenue time.